diff --git a/tester.c b/tester.c
index 9c98901..6f2a0c0 100644
--- a/tester.c
+++ b/tester.c
@@ -45,14 +45,14 @@ void sigint_handler(int sign) {
 }
 
 static void mq_notify_callback(union sigval sv) {
-    memset(buffer, 0, BUF_SIZE);
-    memset(word, 0, BUF_SIZE);
-
     struct mq_attr attr;
 
     bool registered_for_mq_update = false;
 
     do {
+        memset(buffer, 0, BUF_SIZE);
+        memset(word, 0, BUF_SIZE);
+
         if (mq_receive(my_mq_desc, buffer, BUF_SIZE, NULL) < 0) {
             panic_cleanup();
             syserr("Error in rec: ");
